app/scripts/ui.js

```diff
@@ -84,95 +80,35 @@ async function start() {
-   * messages are received thus the background is ready, and ensures that streams and
-   * phishing warning page load only after the "startUISync" message is received.
-   * In case the UI is already rendered, only update the streams.
-   */
-  const messageListener = async (message) => {
-    const method = message?.data?.method;
-
-    switch (method) {
-      case METHOD_START_UI_SYNC:
-        await handleStartUISync();
-        break;
-      case METHOD_DISPLAY_STATE_CORRUPTION_ERROR:
-        handleDisplayStateCorruptionError(message.data.params);
-        break;
-      default:
-    }
-  };
+  const connectionStream = new PortStream(extensionPort);
+  const subStreams = connectSubstreams(connectionStream);
+  const backgroundConnection = metaRPCClientFactory(subStreams.controller);
+  connectToBackground(container, backgroundConnection, handleStartUISync);
+  setupProviderConnection(subStreams.provider);
 
   async function handleStartUISync() {
     endTrace({ name: TraceName.BackgroundConnect });
 
-    if (isManifestV3 && isUIInitialised) {
-      // Currently when service worker is revived we create new streams
-      // in later version we might try to improve it by reviving same streams.
-      updateUiStreams(connectionStream);
-    } else {
-      await initializeUiWithTab(
-        activeTab,
-        connectionStream,
-        windowType,
-        traceContext,
-      );
-    }
+    const activeTab = await queryCurrentActiveTab(windowType);
+
+    await initializeUiWithTab(
+      activeTab,
+      backgroundConnection,
+      windowType,
+      traceContext,
+    );
 
     if (isManifestV3) {
       await loadPhishingWarningPage();
-    } else {
-      extensionPort.onMessage.removeListener(messageListener);
     }
   }
 
-  /**
-   * @typedef {import('./lib/state-corruption-errors').ErrorLike} ErrorLike
-   */
-
-  /**
-   * Updates the DOM with the state corruption error UI.
-   *
-   * @param {{ error: ErrorLike, currentLocale?: string }} params
-   */
-  function handleDisplayStateCorruptionError({ error, currentLocale }) {
-    displayStateCorruptionError(container, error, currentLocale);
-  }
-
-  if (isManifestV3) {
-    // resetExtensionStreamAndListeners takes care to remove listeners from closed streams
-    // it also creates new streams and attaches event listeners to them
-    const resetExtensionStreamAndListeners = () => {
-      extensionPort.onMessage.removeListener(messageListener);
-      extensionPort.onDisconnect.removeListener(
-        resetExtensionStreamAndListeners,
-      );
-
-      extensionPort = browser.runtime.connect({ name: windowType });
-      connectionStream = new PortStream(extensionPort);
-      extensionPort.onMessage.addListener(messageListener);
-      extensionPort.onDisconnect.addListener(resetExtensionStreamAndListeners);
-    };
-
-    extensionPort.onDisconnect.addListener(resetExtensionStreamAndListeners);
```

Comment:
I don't think this does anything. If you force an onDisconnect by doing something like `setTimeout(() => {
   extensionPort.disconnect();
}, 10000);` after doing `extensionPort = browser.runtime.connect({ name: windowType });` the first time (way above), this `resetExtensionStreamAndListeners` listener is never called. Even if it were called the `connectionStream` here is not connected to anything, so messages can't be sent, and received messages would be ignored.